images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/polar_cloud.py

```diff
@@ -0,0 +1,274 @@
+            encrypted = b64encode(cipher_rsa.encrypt(response["challenge"]))
+            logger.debug(f"_on_welcome encrypted challenge: {encrypted}")
+            data = {
+                "serialNumber": self.polar_sn,
+                "signature": encrypted,  # BASE64 encoded string
+                "MAC": self.mac,
+                "protocol": "2.0",
+                "mfgSn": self.serial_number(),
+            }
+            """
+            Note that the following optional fields might be used in future.
+            "printerMake": "printer make",                     // string, optional
+            "version": "currently installed software version", // string, optional
+            "localIP": "printer's local IP address",           // string, optional
+            "rotateImg": 0 | 1,                                // integer, optional
+            "transformImg": 0 - 7,                             // integer, optional
+            "camOff": 0 | 1,                                   // integer, optional
+            "camUrl": "URL for printer's live camera feed"     // string, optional
+            """
+
+            # Send hello request.
+            await self.socket.emit("hello", data)
+        elif not self.polar_sn and not self.public_key:
+            # We need to get an RSA key pair before we can go further.
+            await self.socket.emit("makeKeyPair", {"type": "RSA", "bits": 2048})
+        # elif not self.polar_sn:
+        #     # We already have a key: just register. Technically, there should be
+        #     # no way to get here. Included for completion.
+        #     logger.error(
+        #         "_on_welcome Somehow there are keys with no SN. Reregistering."
+        #     )
+        #     await self._register()
+
+    def _on_hello_response(self, response, *args, **kwargs):
+        if response["status"] == "SUCCESS":
+            logger.info("_on_hello_response success")
+        else:
+            logger.error(f"_on_hello_response failure: {response['message']}")
+            # Deal with error here.
+
+    async def _on_keypair_response(self, response, *args, **kwargs):
+        """
+        Request a keypair from the server. On success register. On failure kick
+        out to interface for new email or pin.
+        """
+        if response["status"] == "SUCCESS":
+            self.public_key = response["public"]
+            self.private_key = response["private"]
+            # We have keys, but still need to register. First disconnect.
+            logger.info("_on_keypair_response success.\nDisconnecting.")
+            # Todo: I'm not creating a race condition with the next three fn calls, am I?
+            await self.socket.disconnect()
+            # After the next line it will send a `welcome`.
+            logger.info("Reconnecting.")
+            await self.socket.connect(self.server_url, transports=["websocket"])
+            await self._register()
```

Comment:
is `connect` synchronous?  (well, as synchronous as `async`/`await` gets?)

Reply:
Yes, `connect` is asynchronous when I use `socketio.AsyncClient()`.